### PR TITLE
Adding cached KVs

### LINKED ISSUE #309: when will kv caching landing?
currently inference is so slow, much more slow than huggingface's transformer implementation, any plan on speed up?

### PR DESCRIPTION
**Pull request by gkroiz**
This PR adds the option to cache KVs for inference. 

Cached KVs portion of splitting https://github.com/Lightning-AI/lit-llama/pull/240 into 2 PRs

main changes:

added KV caching
moved rope cache to the whole model from each attention layer
added mask cache to use for attention masking with cached kvs.
fix ops and tensors that would cause recompilation.

Similar to https://github.com/Lightning-AI/lit-parrot/pull/51
Fixes https://github.com/Lightning-AI/lit-llama/issues/197
Fixes #309

### CODE REVIEW

**Comment by gkroiz** [comment]
TODO: update tests

**Comment by lantiga** [comment]
For visibility: updates to lit-parrot caching
https://github.com/Lightning-AI/lit-parrot/pull/60

File: generate.py [hunk 1]
```diff
@@ -17,10 +17,11 @@
 
 @torch.no_grad()
 def generate(
```
Comment (carmocca) [inline generate.py:19]: The other scripts that call this function also need an update

**Review by carmocca — COMMENTED** [review]
Implementation looks good!

File: generate.py (hunk 1 above)
Comment (gkroiz) [reply generate.py:19]: I'll try to get to this later today

Comment (carmocca) [reply generate.py:19]: I fixed adapter. Need to do a check across the repo to see what's outdated.

Can you update the inference numbers in the tpu howto?

**Comment by gkroiz** [comment]
@carmocca what is the difference between `generate.py` and `generate/full.py`?

**Comment by carmocca** [comment]
@gkroiz It's basically a duplicate. See the description here https://github.com/Lightning-AI/lit-llama/pull/255

**Comment by carmocca** [comment]
I opened https://github.com/Lightning-AI/lit-llama/issues/313 to track it. I wouldn't try to merge them here.

File: generate/full.py [hunk 2]
```diff
@@ -79,7 +25,7 @@ def main(
     top_k: int = 200,
     temperature: float = 0.8,
     checkpoint_path: Optional[Path] = None,
-    tokenizer_path: Path = Path("checkpoints/lit-llama/tokenizer.model"),
+    tokenizer_path: Path = Path("../checkpoints/lit-llama/tokenizer.model"),
```
Comment (carmocca) [inline generate/full.py:28]: The assumption from the howtos is that these scripts are called from the repo root, so `..` shouldn't be used

File: generate.py (hunk 1 above)
Comment (gkroiz) [reply generate.py:19]: Waiting to update these numbers for https://github.com/Lightning-AI/lit-parrot/pull/83#issuecomment-1558042678

**Review by carmocca — APPROVED** [review]
I think we're done here! The howto update can be done separately

**Review by lantiga — APPROVED** [review]
Awesome job!
